Bluetooth: Add `bt_addr_le_eq`

alwa-nordic opened the pull request.

alwa-nordic committed:
Bluetooth: Add `bt_addr_le_eq`

The use of this function is more readable than `bt_addr_cmp`.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jori-nordic commented:
same comment as https://github.com/zephyrproject-rtos/zephyr/pull/50362#issuecomment-1251019936

alwa-nordic committed:
Bluetooth: Replace all '!bt_addr_le_cmp' with 'bt_addr_le_eq'

There should be functional equivalence between these two forms. And the
'_eq'-form is more readable.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Replace 'bt_addr_le_cmp(x) == 0' with 'bt_addr_le_eq(x)'

These should be equivalent. The '_eq'-form is more readable.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Replace `bt_addr_le_cmp` in bool context with `!bt_addr_le_eq`

These should be equivalent. The new form is more readable.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Replace 'bt_addr_le_cmp(x) != 0' with 'bt_addr_le_eq(x)'

These should be equivalent. The '_eq'-form is more readable.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jhedberg commented on include/zephyr/bluetooth/addr.h:
```diff
@@ -81,6 +81,13 @@ static inline int bt_addr_cmp(const bt_addr_t *a, const bt_addr_t *b)
 	return memcmp(a, b, sizeof(*a));
 }
 
+/** @brief Compare Bluetooth LE device addresses for equality.
+ */
```
Since this is a public header file you need to have complete doxygen documentation for the input parameters as well as the return value.

alwa-nordic replied:
Is this really a requirement? I don't think there is any information to add here. Structural equality is a generic operation. In a modern language, it would be derived by the compiler automatically and accessible by a generic (==). More text in the documentation does not make for better documentation. I don't even know what to write in the param documentation that would not make me grimace. "bt_addr_le_t *a: The address to compare equal to `b`".? It just seems redundant.

jhedberg replied:
How does the generated documentation look like if you omit this info?

alwa-nordic replied:
This is how the doc looks like now:
![image](https://user-images.githubusercontent.com/48479610/196637081-ba1feb44-b8cd-482b-a4e4-35954f203cf7.png)

alwa-nordic replied:
I added a `@sa` to `bt_addr_le_cmp`, which now looks like this:
![image](https://user-images.githubusercontent.com/48479610/196637870-266124a3-8f2c-4896-969a-c80d37764cfb.png)

jhedberg commented on include/zephyr/bluetooth/addr.h:
```diff
@@ -81,6 +81,13 @@ static inline int bt_addr_cmp(const bt_addr_t *a, const bt_addr_t *b)
+ */
+static inline bool bt_addr_le_eq(const bt_addr_le_t *a, const bt_addr_le_t *b)
+{
+	return ((a->type == b->type) && (memcmp(a->a.val, b->a.val, sizeof(a->a.val)) == 0));
```
Why not just `return (bt_addr_le_cmp(a, b) == 0);`?

alwa-nordic replied:
I suppose it can be. But I assumed we would like to deprecate `bt_addr_le_cmp`?

Also `bt_addr_le_cmp` is not well defined. The documentation just says "compare" and uses (>)/(<) operators as if it already had a defined meaning on `bt_addr_le_t`. The implementation also just `memcmp`s the whole structure, which is not packed, so probably not meant to be accessed as a string?

jhedberg replied:
The main reason we've had "cmd" functions is to potentially allow them to be used with sorting functions that take a callback with a similar signature (akin to memcmp() and strcmp()). I suppose use cases where you need to sort addresses are quite rare, so we could indeed consider deprecating the cmp function. However if that's part of the intention with these new APIs, please include that information in the commit message. Answering the question "why?" is usually the most important task of the commit message.

alwa-nordic replied:
Ok. I think I was too quick with assuming we want to deprecate `cmp`. I see `cmp` may indeed be useful for algorithms like heaps and sorted log(n) lookups. In that case it does not even matter how the addresses are sorted as long as the transitive (<=) property holds.

I will update this PR with the definition of `bt_addr_le_eq` that you propose.

I will leave the discussion about whether `bt_addr_le_cmp` is well defined for the future. Just as food for thought: My current thinking is that it's best if we follow the rules of C and code for the C abstract machine. So it's not good that `bt_addr_le_cmp` is just a `memcmp`. I believe the compiler is in principle allowed to insert padding as it pleases after any member in a struct. The fully compliant definition of `bt_addr_le_cmp` should be `return (b.type - a.type) || bt_addr_cmp(&a.a, &b.a)`, and `bt_addr_cmp` should be `memcmp(a->val, b->val, sizeof(a->val));`.

jhedberg reviewed (changes requested):
Please add some users for any new APIs you add (can be as separate commits in this same PR).

alwa-nordic committed:
Bluetooth: Add unit tests for bt_addr_le_eq

Add some basic unit tests for bt_addr_le_eq.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jori-nordic commented on include/zephyr/bluetooth/addr.h:
```diff
@@ -87,12 +87,27 @@ static inline int bt_addr_cmp(const bt_addr_t *a, const bt_addr_t *b)
 
+/** @brief Determine equality of two Bluetooth LE device addresses.
+ *
+ *  The Bluetooth LE addresses are equal iff both the types and the 48-bit
```
if
